services/supervisor/src/simple/mod.rs

```diff
@@ -22,17 +22,89 @@ use exonum::{
+///
+/// - The function must be called at the end of the transaction or `before_commit` execution.
+///   If the transaction errors / panics afterwards, the changes to the configs will not
+///   be rolled back.
+/// - No changes to the blockchain state should be introduced before the call to this function.
+///   Any changes that are introduced will be committed regardless of the execution status,
+///   or the status of application of any config change. This is if the execution wasn't interrupted
+///   by a panic / error *before* hitting the call; if this happens, the usual rules apply.
+///
+/// These restrictions are the result of `Fork` not having multi-layered checkpoints.
+fn update_configs(context: &mut CallContext, changes: Vec<ConfigChange>) {
+    // An error while configuring one of the service instances should not affect others.
+    changes.into_iter().for_each(|change| match change {
+        ConfigChange::Consensus(config) => {
+            log::trace!("Updating consensus configuration {:?}", config);
+            let configure_result = apply_consensus_config(context, config);
+
+            if let Err(e) = configure_result {
+                // Consensus configuration is not allowed to fail, since it was validated and approved.
+                // This is an unrecoverable error.
+                panic!("Consensus configuration failed with error {:?}", e)
+            }
+        }
+
+        ConfigChange::Service(config) => {
+            log::trace!(
+                "Updating service instance configuration, instance ID is {}",
+                config.instance_id
+            );
+            let configure_result = apply_service_config(context, config);
+
+            if let Err(e) = configure_result {
+                log::error!(
+                    "An error occurred while applying service configuration. {}",
+                    e
+                );
+            }
+        }
+
+        ConfigChange::DeployRequest(deploy_request) => {
+            log::trace!("Processing deploy request {:?}", &deploy_request);
+            // Activation height is ignored for simple supervisor.
+            let start_deploy_result =
+                context.start_artifact_registration(deploy_request.artifact, deploy_request.spec);
+            if let Err(e) = start_deploy_result {
+                // Deployment failures are not allowed within simple supervisor.
+                panic!("Deployment failed with error {:?}", e);
+            }
+        }
+
+        ConfigChange::StartService(start_request) => {
+            log::trace!("Processing start service request {:?}", &start_request);
+            let start_instance_result = context.start_adding_service(
+                start_request.artifact,
+                start_request.name,
+                start_request.config,
+            );
+            if let Err(e) = start_instance_result {
+                // Service start failures are not allowed within simple supervisor.
```

Comment:
Well, since we don't use confirmations for deployment, I'm not sure how we can distinguish if service deployment failed for one node or for the whole network.

Regarding panics: again, I'm not sure what'll be the best solution here (as well as I'm not sure that service should be considered still running after panicking, even though we don't support service stopping yet).

Reply:
> I'm not sure how we can distinguish if service deployment failed for one node or for the whole network.

Via consensus algorithm, ideally. I think you're implying that the dispatcher schema should become merkelized again 🙃 (Which seems possible to do if it won't distinguish between active and pending services / artifacts.)

Reply:
Yep, I guess that's one of the options. Anyway, the responsibility here is not on the `SimpleSupervisor` in any scenario (I believe), so **for the `SimpleSupervisor`** (without any assumptions about consensus or internal node algorithms) this case is a critical error.